Do not try to create bitmap for default constructed Caret

StefanBruens opened the pull request:
If a Caret is default-constructed, the width and height are still 0,
and the wxBitmap::Create implementations expect valid sizes.

StefanBruens committed:
Do not try to create bitmap for default constructed Caret

If a Caret is default-constructed, the width and height are still 0,
and the wxBitmap::Create implementations expect valid sizes.

The same applies for the bitmap after a size change.

vadz commented:
Thanks, there is clearly a problem here, but I'm not sure this is enough to solve it, as it looks like the bitmap will never be created for a default constructed object and we should override `DoCreate()` to do it. Or am I missing something here?

vadz commented:
Sorry, I don't understand how does this answer my question, so perhaps I didn't ask the question clearly enough. Consider what happens with
```
wxCaret c;
c.Create(win, 10, 10);
```
Previously this created an invalid bitmap for `m_bmpUnderCaret`. Now this doesn't create any bitmap at all, i.e. `m_bmpUnderCaret` remains uninitialized. This is not really an improvement, i.e. default-constructed `wxCaret` still doesn't work.

I think we should fix.

But, looking at this code again, perhaps an even better way to do it would be just dropping `wxHAS_CARET_USING_OVERLAYS` and always using `wxOverlay` for it, as I think the generic `wxOverlay` implementation should be good enough for this.

StefanBruens commented:
> Sorry, I don't understand how does this answer my question, so perhaps I didn't ask the question clearly enough. Consider what happens with
> 
> ```
> wxCaret c;
> c.Create(win, 10, 10);
> ```

m_bmpUnderCaret is default-constructed, so it is initialized (m_refData == NULL).

Due to `wxCHECK_MSG`, `wxBitmap::Create(0, 0, x)` has no effect at all, save the error message:

https://github.com/wxWidgets/wxWidgets/blob/cc931612eec2e3ea49200ebff45042135f3c3f9c/src/gtk/bitmap.cpp#L537-L541

vadz commented:
OK, I see the source of misunderstanding: `wxBitmap` (as several other similar classes) is (smart)pointer-like, so while it is initialized, it's in an "empty" state, which is what I meant by "not initialized", as it can't be used in this state.

Adding `SetSize()` would work, but it shouldn't be needed, in wxAPI using default ctor and calling `Create(args...)` must have exactly the same effect as using ctor and passing it `args...`.

vadz commented:
Yes, exactly. And while fixing the warning is nice, the actual problem is that using `Create()` just does not work and it would be really better to fix this.

StefanBruens committed:
Create underlying bitmap for caret when it is explicitly created

wxCaret::Create implies a size change, so call DoSize to initialize
the underlying wxBitmap.

StefanBruens commented:
@vadz Ping!

vadz commented:
Sorry, I had to spend all my time on catching up on other projects and didn't have any time for wx recently. I'll have a look at this a.s.a.p.

vadz commented:
Thanks for the fixes and sorry for the very long delay (OTOH there is a pleasant regularity here: your first commit was done on 2020-07-31, the second one on 2020-08-31 and now it has been merged on basically 2020-09-31 :-)! I've just added a unit test and fixed it now.
